### ISSUE TITLE
GitHub Pages: `quarto publish gh-pages` stalls at "deploying gh-pages ..." for _private_ GitHub Pages

### ISSUE THREAD

**Issue by HenrikBengtsson** [quarto-dev/quarto-cli#2864]
### Bug description

# Issue

When using `quarto publish gh-pages` to push to _private_ GitHub Pages, the tool get stuck on "Deploying gh-pages branch to website", despite the updated GitHub Pages is deployed and live.

# Example

```sh
$ quarto publish gh-pages
? Publish site to https://example-org.github.io/quarto-book-test/ using gh-pages? (Y/n) › Yes
Switched to a new branch 'gh-pages'
[gh-pages (root-commit) b53766a] Initializing gh-pages branch
...
Preparing worktree (resetting branch 'gh-pages'; was at b53766a)
Branch 'gh-pages' set up to track remote branch 'gh-pages' from 'origin'.
HEAD is now at b53766a Initializing gh-pages branch
[gh-pages dd3779a] Built site for gh-pages
 23 files changed, 5334 insertions(+)
...
origin	git@github.com:example-org/quarto-book-test.git (fetch)
origin	git@github.com:example-org/quarto-book-test.git (push)
X11 forwarding request failed on channel 0
To github.com:example-org/quarto-book-test.git
   b53766a..dd3779a  HEAD -> gh-pages

(/) Deploying gh-pages branch to website (this may take a few minutes)
```

The latter progress updater never completes. I had to hit <kbd>Ctrl-C</kbd>.

# Troubleshooting

My best guess is that the tool is polling the wrong URL, because _private_ GitHub Pages uses a "random" `*.pages.github.io` URL for security reasons, e.g. <https://upgraded-spork-34abee09.pages.github.io/>. This is explained at <https://docs.github.com/en/enterprise-cloud@latest/pages/getting-started-with-github-pages/changing-the-visibility-of-your-github-pages-site#about-subdomains-for-privately-published-sites>.

# Session info

This is on Ubuntu 20.04 running:

```sh
$ quarto --version
1.2.215
```

I'm pushing to a _private_ repository configured to use _private_ GitHub Pages.  This is hosted under a GitHub Enterprise Cloud plan.

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [ ] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by jjallaire** [comment]
Yes, we are indeed polling the public URL! Let me see if we get a distinct HTTP status for private repos which would let us know to stop polling.
We could also just poll the known public endpoints for the repo and if they 404 then give up on the public codepath. Should be able to resolve this in short order.

**Comment by HenrikBengtsson** [comment]
And/or maybe an CLI option `--no-wait`, `--push-only`, or something like that.

**Comment by HenrikBengtsson** [comment]
Ah, I should have checked - it looks like

```sh
$ quarto publish gh-pages --no-browser
...
To github.com:example-org/quarto-book-test.git
   36d10f4..26b1742  HEAD -> gh-pages

NOTE: GitHub Pages sites use caching so you might need to click the refresh
button within your web browser to see changes after deployment.

[✓] Published to https://example-org.github.io/quarto-book-test/

NOTE: GitHub Pages deployments normally take a few minutes (your site updates
will be visible once the deploy completes)
```

skips the polling.

**Comment by HenrikBengtsson** [comment]
> Let me see if we get a distinct HTTP status for private repos which would let us know to stop polling.
> We could also just poll the known public endpoints for the repo and if they 404 then give up on the public codepath. Should be able to resolve this in short order.

It's an HTTP/2 301 redirect;

```sh
$ curl --head https://example-org.github.io/quarto-book-test/
HTTP/2 301 
server: GitHub.com
content-type: text/html
permissions-policy: interest-cohort=()
location:  https://upgraded-spork-34abee09.pages.github.io/
x-github-request-id: 8201:2F3F:C1010:D3B61:6ECE349B
accept-ranges: bytes
date: Fri, 14 Oct 2022 23:12:43 GMT
via: 1.1 varnish
age: 0
x-served-by: cache-sjc10036-SJC
x-cache: MISS
x-cache-hits: 0
x-timer: S1665789163.230189,VS0,VE103
vary: Accept-Encoding
x-fastly-request-id: 4c058c9dd8688e01b7cc90ee1c135f5e7f44e13a
content-length: 162
```

**Comment by jjallaire** [comment]
Reliably figuring out when a private site is in play has proven a tad trickier than I'd hoped. We are just about to release v1.2 and I don't want to make a change too quickly that might break the main flow. In the meantime I've updated the docs to recommend `--no-browser` for the private site scenario: https://quarto.org/docs/publishing/github-pages.html#publishing

**Comment by HenrikBengtsson** [comment]
Sounds good. FYI, there's a stray "Private Sites" sentence in <https://quarto.org/docs/publishing/github-pages.html#publishing> just before the 'Private Sites' subsection.

**Comment by jjallaire** [comment]
Thanks! Just fixed.

**Comment by ludwigwittgenstein2** [comment]
**I am facing the same issue**
**I am stuck, any suggestions?**


32 files changed, 5565 insertions(+)
 create mode 100644 c6c48ed8/.nojekyll
 create mode 100644 c6c48ed8/about.html
 create mode 100644 c6c48ed8/index.html
 create mode 100644 c6c48ed8/listings.json
 create mode 100644 c6c48ed8/posts/post-with-code/image.jpg
 create mode 100644 c6c48ed8/posts/post-with-code/index.html
 create mode 100644 c6c48ed8/posts/welcome/index.html
 create mode 100644 c6c48ed8/posts/welcome/thumbnail.jpg
 create mode 100644 c6c48ed8/posts/welcome/thumbnail.png
 create mode 100644 c6c48ed8/profile.jpg
 create mode 100644 c6c48ed8/robots.txt
 create mode 100644 c6c48ed8/search.json
 create mode 100644 c6c48ed8/site_libs/bootstrap/bootstrap-icons.css
 create mode 100644 c6c48ed8/site_libs/bootstrap/bootstrap-icons.woff
 create mode 100644 c6c48ed8/site_libs/bootstrap/bootstrap.min.css
 create mode 100644 c6c48ed8/site_libs/bootstrap/bootstrap.min.js
 create mode 100644 c6c48ed8/site_libs/clipboard/clipboard.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-html/anchor.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-html/popper.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-html/quarto-syntax-highlighting.css
 create mode 100644 c6c48ed8/site_libs/quarto-html/quarto.js
 create mode 100644 c6c48ed8/site_libs/quarto-html/tippy.css
 create mode 100644 c6c48ed8/site_libs/quarto-html/tippy.umd.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-listing/list.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-listing/quarto-listing.js
 create mode 100644 c6c48ed8/site_libs/quarto-nav/headroom.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-nav/quarto-nav.js
 create mode 100644 c6c48ed8/site_libs/quarto-search/autocomplete.umd.js
 create mode 100644 c6c48ed8/site_libs/quarto-search/fuse.min.js
 create mode 100644 c6c48ed8/site_libs/quarto-search/quarto-search.js
 create mode 100644 c6c48ed8/sitemap.xml
 create mode 100644 c6c48ed8/styles.css
origin  https://github.com/ludwigwittgenstein2/Tamil.git (fetch)
origin  https://github.com/ludwigwittgenstein2/Tamil.git (push)
To https://github.com/ludwigwittgenstein2/Tamil.git
 + 8634b58...72ca7b6 HEAD -> gh-pages (forced update)
fatal: 'c6c48ed8' is not a working tree

NOTE: GitHub Pages sites use caching so you might need to click the refresh
button within your web browser to see changes after deployment.

(/) Deploying gh-pages branch to website (this may take a few minutes)
